src/lib.rs

```diff
@@ -102,60 +105,120 @@ struct Reaper {
 
     /// The pipe that delivers signal notifications.
     pipe: Pipe,
+
+    /// Locking this mutex indicates that we are polling the SIGCHLD event.
+    driver_guard: AsyncMutex<()>,
+
+    /// The number of tasks polling the SIGCHLD event.
+    ///
+    /// If this is zero, the `async-process` thread must be spawned.
+    drivers: AtomicUsize,
+
+    /// Ensures the `async-process` thread is only ever spawned once.
+    async_process_thread: Once,
+
+    /// Number of live `Child` instances currently running.
+    ///
+    /// This is used to prevent the reaper thread from being spawned right as the program closes,
+    /// when the reaper thread isn't needed. This represents the number of active processes.
+    child_count: AtomicUsize,
 }
 
 impl Reaper {
     /// Get the singleton instance of the reaper.
     fn get() -> &'static Self {
         static REAPER: OnceCell<Reaper> = OnceCell::new();
 
-        REAPER.get_or_init_blocking(|| {
+        REAPER.get_or_init_blocking(|| Reaper {
+            sigchld: Event::new(),
+            zombies: Mutex::new(Vec::new()),
+            pipe: Pipe::new().expect("cannot create SIGCHLD pipe"),
+            driver_guard: AsyncMutex::new(()),
+            drivers: AtomicUsize::new(0),
+            async_process_thread: Once::new(),
+            child_count: AtomicUsize::new(0),
+        })
+    }
+
+    /// Ensure that the reaper is driven.
+    ///
+    /// If there are no active `driver()` callers, this will spawn the `async-process` thread.
+    #[inline]
+    fn ensure_driven(&'static self) {
+        if self.drivers.load(Ordering::Acquire) == 0 {
+            self.start_driver_thread();
+        }
+    }
+
+    /// Start the `async-process` thread.
+    #[cold]
+    fn start_driver_thread(&'static self) {
+        self.async_process_thread.call_once(move || {
```

Comment:
this `Once` thing here irritates me a bit... because this is only called by `ensure_driven` afaik, and we check the `self.drivers` count before that (it might make more sense to do CAS there to ensure that only one thread tries to spawn the reaper async-process, and don't increment the drivers count in `start_driver_thread`)

Reply:
I think the `Once` is necessary in case another thread calls `Child::new` while the thread is being spawned.

Reply:
well, but that would also be solved by handling the `ensure_driven` with a bit more "leverage" (that is, using CAS to negotiate who spawns the thread). .... The only thing that might be a bit more difficult is ensuring that the resulting code is panic safe (so a panic during this spawn process doesn't prevent the driver thread from being spawned later)

Reply:
in fact, I would appreciate it if the `start_driver_thread` method would be inlined into `ensure_driven` (e.g. starting the method with the following:)
```rust
        if self.drivers.compare_exchange(0, 1, Ordering::SeqCst, Ordering::Acquire).is_err() {
            return;
        }
        // inlined self.start_driver_thread()
```